src/Common/SIEVECachePolicy.h

```diff
@@ -0,0 +1,212 @@
+
+        Cell & cell = it->second;
+        cell.visited = cell.visited >= up ? cell.visited : cell.visited + 1;
+        /// Move the key to the end of the queue. The iterator remains valid.
+        // queue.splice(queue.end(), queue, cell.queue_iterator);
+
+        return cell.value;
+    }
+
+    std::optional<KeyMapped> getWithKey(const Key & key) override
+    {
+        auto it = cells.find(key);
+        if (it == cells.end())
+            return std::nullopt;
+
+        Cell & cell = it->second;
+        cell.visited = cell.visited >= up ? cell.visited : cell.visited + 1;
+        /// Move the key to the end of the queue. The iterator remains valid.
+        // queue.splice(queue.end(), queue, cell.queue_iterator);
+
+        return std::make_optional<KeyMapped>({it->first, cell.value});
+    }
+
+    void set(const Key & key, const MappedPtr & mapped) override
+    {
+        auto [it, inserted] = cells.emplace(std::piecewise_construct,
+            std::forward_as_tuple(key),
+            std::forward_as_tuple());
+
+        Cell & cell = it->second;
+
+        if (inserted)
+        {
+            try
+            {
+                cell.queue_iterator = queue.insert(queue.end(), key);
+                cell.visited = 0;
+            }
+            catch (...)
+            {
+                cells.erase(it);
+                throw;
+            }
+        }
+        else
+        {
+            current_size_in_bytes -= cell.size;
+            cell.visited = cell.visited >= up ? cell.visited : cell.visited + 1;
+            // queue.splice(queue.end(), queue, cell.queue_iterator);
+        }
+
+        cell.value = mapped;
+        cell.size = cell.value ? weight_function(*cell.value) : 0;
+        current_size_in_bytes += cell.size;
+
+        removeOverflow();
+    }
+
+    std::vector<KeyMapped> dump() const override
+    {
+        std::vector<KeyMapped> res;
+        for (const auto & [key, cell] : cells)
+            res.push_back({key, cell.value});
+        return res;
+    }
+
+private:
+    using SIEVEQueue = std::list<Key>;
+    using SIEVEQueueIterator = typename SIEVEQueue::iterator;
+
+    SIEVEQueue queue;
+    typename SIEVEQueue::iterator position;
+    const size_t up = 1;
+    struct Cell
+    {
+        MappedPtr value;
+        size_t size;
+        size_t visited;
+        SIEVEQueueIterator queue_iterator;
+    };
+
+    using Cells = std::unordered_map<Key, Cell, HashFunction>;
+
+    Cells cells;
+
+    /// Total weight of values.
+    size_t current_size_in_bytes = 0;
+    size_t max_size_in_bytes;
+    size_t max_count;
+
+    WeightFunction weight_function;
+    OnWeightLossFunction on_weight_loss_function;
+
+    void removeOverflow()
+    {
+        size_t current_weight_lost = 0;
+        size_t queue_size = cells.size();
+
+        while ((current_size_in_bytes > max_size_in_bytes || (max_count != 0 && queue_size > max_count)) && (queue_size > 0))
+        {
+            const Key & key = *position;
```

Comment:
This code does not work correctly, we should check if position points to a cell with `visited = 0`, if yes - evict, otherwise set `visited = 0` and move position further and proceed.

Reply:
I do not see this comment being resolved as well, please do not resolve the comments unless fixed.

Reply:
This is my first time submitting a PR, and there are some things I'm not very familiar with. I apologize for any inconvenience my previous mistakes may have caused.